### PR TITLE
interpolation: fix mirror and wrap border modes

### PR DESCRIPTION
**Pull request by dtorop**
Mirror math was off such that could reference memory below the minimum of the clip range. Mirror mode is used when computing single samples, which is used by various iops (e.g. lens, liquify, ashift, scalepixels, rotatepixels), and hence could cause an unexpected memory reference, perhaps including one before the start of the image.

Wrap math was off such that when wrapping from min to max, it would actually wrap to max-1 (missing max). And conversely when wrapping from max to min, it would wrap to min + 1 (missing min). Wrap code is unused, so this should have had no effect.

Also, some minor fixups:

- Clarify comments describing dt_interpolation_resample*() functions: output runs to roi width/height less 1
- Tidy indentation
- Update copyright lines

I'm still going through the code to figure out the offset problems between preview/full pixelpipes. This PR came from looking carefully through relevant code, but is orthogonal to that work, hence is a separate PR.

### CODE REVIEW

File: src/common/interpolation.c [hunk 1]
```diff
@@ -107,7 +107,7 @@ static inline int clip(int i, int min, int max, enum border_mode mode)
     case BORDER_MIRROR:
       if(i < min)
       {
-        i = min - i;
+        i = 2 * min - i;
```
Comment (jenshannoschwalm) [inline src/common/interpolation.c:110]: It might be good to add an explaining comment here why we need `2*min` , it took me a while to understand this :-)

**Comment by jenshannoschwalm** [comment]
Nice catch, could indeed a read from valid_address-1

**Comment by jenshannoschwalm** [comment]
Wouldn't this be also for 4.2.1 ?

File: src/common/interpolation.c (hunk 1 above)
Comment (TurboGit) [reply src/common/interpolation.c:110]: Yes, I'm not sure to fully understand this.

**Comment by TurboGit** [comment]
> Wouldn't this be also for 4.2.1 ?

I'm not sure... Have we got issues reported for this? In a sense killing a bug in a bug-fix release is good, but as I'm not yet sure to properly understand the patch (and the possible implication/side-effects) I'm a bit nervous at the idea to introduce this in 4.2.1. But I'll listen to both of you are certainly more expert than me on this part.

**Comment by dtorop** [comment]
Interestingly enough: If `min` is 0, then the code as it stands does the right thing! That is, in the special case of `min = 0`, `2*min - i` == `min - i`! And as all calls to `clip()` set `min` to `0`, in the case of the current code, `clip()` works flawlessly. Hence I'd suggest still fixing the code, but to mark this as for 4.4. 

To try to interpret this:
- The `BORDER_MIRROR` implementation is wrong, but in all its use cases it won't fail.
- It is _possible_ that the `BORDER_MIRROR` implementation is extremely cleverly optimized for its use cases. But in that case it would still make sense to clarify its parameters (e.g. drop the `min` parameter if we know it's always 0, and adjust accordingly, e.g. optimize further to a simple negation).
- The compiler is capable enough to know that `clip()` is being inlined with 0 as `min`, and change the math accordingly. So there's no reason not to simply make the code correct.
- My guess is that this all came from a simple copy/paste error, but as `min` is always 0, the result was inconsequential. (Unless somewhere ML program is reading through this codebase and will one day output a flawed `clip()` function to a gullible user.)
- Regardless, this isn't a huge deal, as in the current code (and most conceivable code), we're not going to be making bad memory accesses.

Here's a godbolt link for working this all out: https://godbolt.org/z/718MbYzME.

It should show that:

In the case of a non-zero min, the current code (results in right column, "old") will produce flawed results. E.g. with `min` of 5 and `max` at 10:

```
     verbose  compact    old
 1 ->      9       9       4
 2 ->      8       8       3
 3 ->      7       7       2
 4 ->      6       6       1
----------------------------
 5 ->      5       5       5
 6 ->      6       6       6
 7 ->      7       7       7
 8 ->      8       8       8
 9 ->      9       9       9
10 ->     10      10      10
----------------------------
11 ->      9       9       9
12 ->      8       8       8
13 ->      7       7       7
14 ->      6       6       6
```

But in the case of `min` of 0, all versions of the code work fine:

```
     verbose    min0     old
-4 ->      4       4       4
-3 ->      3       3       3
-2 ->      2       2       2
-1 ->      1       1       1
----------------------------
 0 ->      0       0       0
 1 ->      1       1       1
 2 ->      2       2       2
 3 ->      3       3       3
 4 ->      4       4       4
 5 ->      5       5       5
----------------------------
 6 ->      4       4       4
 7 ->      3       3       3
 8 ->      2       2       2
 9 ->      1       1       1
```

And that rejiggering the math as in my most recent commit still produces pretty much the same assembly (here is what clang produces, mirror_verbose = most recent commit in this PR, mirror_compact is from prior commit):

```asm
mirror_verbose:                         # @mirror_verbose
        mov     eax, edi
        cmp     esi, edi
        jle     .LBB0_2
        add     esi, esi
        sub     esi, eax
        mov     eax, esi
        ret
.LBB0_2:
        cmp     edx, eax
        jge     .LBB0_4
        add     edx, edx
        sub     edx, eax
        mov     eax, edx
.LBB0_4:
        ret

mirror_compact:                         # @mirror_compact
        mov     eax, edi
        cmp     edi, esi
        jge     .LBB1_2
        add     esi, esi
        sub     esi, eax
        mov     eax, esi
        ret
.LBB1_2:
        cmp     eax, edx
        jle     .LBB1_4
        add     edx, edx
        sub     edx, eax
        mov     eax, edx
.LBB1_4:
        ret
```

And that in the case of the inlined/optimized variant with min set to 0, the assembly is reasonable (clang variant again, function mirror_min0) and not that far off from the code output by current dt codebase (mirror_old):

```asm
mirror_min0:                            # @mirror_min0
        mov     eax, edi
        test    edi, edi
        js      .LBB0_1
        cmp     esi, eax
        jge     .LBB0_4
        add     esi, esi
        sub     esi, eax
        mov     eax, esi
.LBB0_4:
        ret
.LBB0_1:
        neg     eax
        ret


mirror_old:                             # @mirror_old
        mov     eax, esi
        sub     eax, edi
        jg      .LBB2_4
        cmp     edi, edx
        jle     .LBB2_2
        add     edx, edx
        sub     edx, edi
        mov     eax, edx
.LBB2_4:
        ret
.LBB2_2:
        mov     eax, edi
        ret
```

I have to say I'm intrigued by the gcc output, which has one fewer branch:

```asm
mirror_min0:
        test    edi, edi
        js      .L6
        mov     edx, edi
        mov     eax, esi
        sub     edx, esi
        sub     eax, edx
        cmp     edi, esi
        cmovle  eax, edi
        ret
.L6:
        mov     eax, edi
        neg     eax
        ret

mirror_compact:
        cmp     edi, esi
        jl      .L11
        lea     eax, [rdx+rdx]
        sub     eax, edi
        cmp     edi, edx
        cmovle  eax, edi
        ret
.L11:
        lea     eax, [rsi+rsi]
        sub     eax, edi
        ret
```

It  appears to do the subtraction no matter what, then only use the result if needed? But gcc developers probably knows a thing or two abut CPU opcode execution...

File: src/common/interpolation.c (hunk 1 above)
Comment (dtorop) [reply src/common/interpolation.c]: Added a commit with better comments. In process of doing that, realized that this is probably a much lower priority bug. See note below.

**Comment by TurboGit** [comment]
@dtorop : Thanks for all the details, very interesting to see such a deep analyze for what could seem a simple arithmetic on darktable C code.

**Review by TurboGit — APPROVED** [review]
Thanks!

**Comment by dtorop** [comment]
Yes! Still more work to do on the offset images. I'm going to be a bit swamped for a few weeks, but will hope to have some results in a bit...

**Comment by TurboGit** [comment]
Entry for the release notes needed please. Thanks.

**Comment by TurboGit** [comment]
Thanks!

**Pull request merged**